driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +191,154 @@ public int getMaxWireVersion() {
+                                    } else {
+                                        wrappedCallback.onResult(nextBatch, null);
+                                    }
+                                }), callback);
     }
 
-    private void getMore(final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
-        if (pinnedConnection != null)  {
-            getMore(pinnedConnection.retain(), cursor, callback);
-        } else {
-            connectionSource.getConnection((connection, t) -> {
-                if (t != null) {
-                    endOperationInProgress();
-                    callback.onResult(null, t);
-                } else {
-                    getMore(assertNotNull(connection), cursor, callback);
-                }
-            });
-        }
+    private CommandCursorResult<T> toCommandCursorResult(final ServerAddress serverAddress, final String fieldNameContainingBatch,
+            final BsonDocument commandCursorDocument) {
+        CommandCursorResult<T> commandCursorResult = getCommandCursorResult(serverAddress, fieldNameContainingBatch, commandCursorDocument);
+        this.count.addAndGet(commandCursorResult.getResults().size());
+        return commandCursorResult;
     }
 
-    private void getMore(final AsyncConnection connection, final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
-        connection.commandAsync(namespace.getDatabaseName(), asGetMoreCommandDocument(cursor.getId(), connection.getDescription()),
-                NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(), CommandResultDocumentCodec.create(decoder, "nextBatch"),
-                connectionSource.getOperationContext(), new CommandResultSingleResultCallback(connection, cursor, callback));
-   }
+    private boolean limitReached() {
+        return Math.abs(limit) != 0 && count.get() >= Math.abs(limit);
+    }
 
-    private BsonDocument asGetMoreCommandDocument(final long cursorId, final ConnectionDescription connectionDescription) {
-        BsonDocument document = new BsonDocument("getMore", new BsonInt64(cursorId))
-                .append("collection", new BsonString(namespace.getCollectionName()));
+    @ThreadSafe
+    private static final class ResourceManager extends CursorResourceManager<AsyncConnectionSource, AsyncConnection> {
 
-        int batchSizeForGetMoreCommand = Math.abs(getNumberToReturn(limit, this.batchSize, count.get()));
-        if (batchSizeForGetMoreCommand != 0) {
-            document.append("batchSize", new BsonInt32(batchSizeForGetMoreCommand));
+        ResourceManager(
+                final MongoNamespace namespace,
+                final AsyncConnectionSource connectionSource,
+                @Nullable final AsyncConnection connectionToPin,
+                @Nullable final ServerCursor serverCursor) {
+            super(new ReentrantLock(), namespace, connectionSource, connectionToPin, serverCursor);
```

Comment:
I'm not 100% sure why two different locks were used / favored originally.

@stIncMale any insights?

Reply:
I used to think that choosing between `ReentrantLock` and `StampedLock.asWriteLock` <sup>1</sup> is trading reentrancy + optional fairness for lower overhead + the ability to unlock from an arbitrary thread (this one may be necessary in asynchronous code in rare situations), so I was defaulting to `StampedLock.asWriteLock` unless I need either reentrancy or fairness.

Turned out (I measured with OpenJDK JDK 20), `ReentrantLock` and `StampedLock.asWriteLock` have virtually identical throughput both when the critical section is empty and when it involves a bit of computations. Thus, actually, the trade is between reentrancy + optional fairness and the ability to unlock from an arbitrary thread. Since the former is needed much more often, I no longer default to `StampedLock.asWriteLock`, but there is still code where I used it just because I defaulted to it (like `QueryBatchCursor`). There is nothing wrong with that, but there is no reason to preserve `StampedLock.asWriteLock` if it introduces any inconvenience.

---

<sup>1</sup> `StampedLock` is most famous for its optimistic reads functionality, which involves no locking, but this is irrelevant to us here.

Reply:
Updated to use RentrantLock in the base class.